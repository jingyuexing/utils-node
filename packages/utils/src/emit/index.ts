export function useEmit() {
   const emitList = new Map<string, Function[]>();
   return {
      Emit<T extends string>(name: T) {
         const list = emitList.get(name) ?? [];
         return function (target: any, propertyName: string) {
            list.push(target[propertyName]);
            emitList.set(name, list);
         };
      },
      Handler<T extends string>(name: T, ...args: any[]): Utils.Dict<T, any[]> {
         let events = emitList.get(name);
         let obj: any = {};
         if (events) {
            obj[name] = events.map(func => {
               return func(args);
            });
            return obj;
         } else {
            obj[name] = [];
            return obj;
         }
      },
   };
}
