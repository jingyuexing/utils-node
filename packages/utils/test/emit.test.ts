import { describe, expect, it } from 'vitest';
import { useEmit } from '@/emit';
describe('test emit', () => {
   let { Emit, Handler } = useEmit();
   it('test emit handler', () => {
      class User {
         @Emit('user')
         name(...args: any[]) {
            console.log('hello');
            return args;
         }
      }
      let { user } = Handler('user', 1, 2, 3, 4, 5, 6);
      expect(user).toMatchInlineSnapshot(`
          [
            [
              [
                1,
                2,
                3,
                4,
                5,
                6,
              ],
            ],
          ]
        `);
   });
});
