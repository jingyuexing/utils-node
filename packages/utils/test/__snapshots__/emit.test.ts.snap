// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test emit > test emit handler > some things is wrong 1`] = `
[
  [
    [
      1,
      2,
      3,
      4,
      5,
      6,
    ],
  ],
]
`;
